pkg/commands/credentials.go

```diff
@@ -0,0 +1,299 @@
+package commands
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"net/rpc"
+	"os"
+	"os/exec"
+	"regexp"
+	"strings"
+
+	"github.com/mgutz/str"
+	"github.com/mjarkk/go-ps"
+)
+
+// Listener is the the type that handles is the callback for server responses
+type Listener int
+
+// InputQuestion is what is send by the lazygit client
+type InputQuestion struct {
+	PublicKey string
+	Question  string
+	Listener  string
+}
+
+// listenerMetaType is a listener there private key and ask function
+type listenerMetaType struct {
+	AskFunction func(string) string
+	AskedFor    struct {
+		Password bool
+		Username bool
+	}
+}
+
+var listenerMeta = map[string]listenerMetaType{} // a list of listeners
+var totalListener uint32                         // this gets used to set the key of listenerMeta
+
+// Input interacts with the lazygit client spawned by git
+func (l *Listener) Input(in InputQuestion, out *EncryptedMessage) error {
+	suspiciousErr := errors.New("closing message due to suspicious behavior")
+
+	listener, ok := listenerMeta[in.Listener]
+	if !ok {
+		return suspiciousErr
+	}
+
+	if !HasLGAsSubProcess() {
+		return suspiciousErr
+	}
+
+	updateListenerMeta := func() {
+		listenerMeta[in.Listener] = listener
+	}
+
+	updateListenerMeta()
+
+	question := in.Question
+
+	prompts := map[string]string{
+		"password": `Password\s*for\s*'.+':`,
+		"username": `Username\s*for\s*'.+':`,
+	}
+
+	var toSend string
+
+	for askFor, pattern := range prompts {
+		match, _ := regexp.MatchString(pattern, question)
+		if match && ((askFor == "password" && !listener.AskedFor.Password) || (askFor == "username" && !listener.AskedFor.Username)) {
```

Comment:
I think we can clean this part up a little bit by making our `prompts` variable be a map from string to struct rather than string to string, and in the struct we can have a Pattern, and an AskedFor pointer like so:

```
	prompts := map[string]prompt{
		"password": prompt{pattern: `Password\s*for\s*'.+':`, askedFor: &listener.AskedFor.Password},
		"username": prompt{pattern: `Username\s*for\s*'.+':`, askedFor: &listener.AskedFor.Username},
	}
```

Then you can say 
```
if match && !*prompt.askedFor {
    *prompt.askedFor = true
    updateListenerMeta()
    ...
}
```

Reply:
Thanks that seems like a better thing to do here